### PR TITLE
Update chain() and cross_downstream() to support XComArgs

### LINKED ISSUE #16635: Update `airflow.models.baseoperator.chain()` function to support XComArgs
**Description**
The `airflow.models.baseoperator.chain()` is a very useful and convenient way to add sequential task dependencies in DAGs but the function only supports tasks of a `BaseOperator` type.

**Use case / motivation**
Users who create tasks via the `@task` decorator will not be able to use the `chain()` function to apply sequential dependencies that do not share an `XComArg` implicit dependency.

**Are you willing to submit a PR?**
Absolutely. 🚀 

**Related Issues**
None

### PR DESCRIPTION
**Pull request by josh-fell**
- Added support for `XComArgs` in `chain()` helper function
- Updated type annotations for `XComArg` in `cross_downstream()` helper function
- Added additional tests in `test_baseoperator.py`

Closes: #16635

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/models/baseoperator.py [hunk 1]
```diff
@@ -1566,36 +1567,41 @@ def chain(*tasks: Union[BaseOperator, Sequence[BaseOperator]]):
         t4.set_downstream(t6)
         t5.set_downstream(t6)
 
-    :param tasks: List of tasks or List[airflow.models.BaseOperator] to set dependencies
-    :type tasks: List[airflow.models.BaseOperator] or airflow.models.BaseOperator
+    :param tasks: List of tasks, List[airflow.models.BaseOperator], XComArg, or List[airflow.models.XComArg]
+        to set dependencies
```
Comment (ashb) [inline airflow/models/baseoperator.py]: As we are changing this line anyway, we don't need to duplicate the types here -- prose is better

```suggestion
    :param tasks: List of tasks or XComArgs to set dependencies
```

File: airflow/models/baseoperator.py [hunk 2]
```diff
@@ -1566,36 +1567,41 @@ def chain(*tasks: Union[BaseOperator, Sequence[BaseOperator]]):
...
             up_task.set_downstream(down_task)
             continue
-        if isinstance(down_task, BaseOperator):
+        if isinstance(down_task, (BaseOperator, XComArg)):
             down_task.set_upstream(up_task)
             continue
         if not isinstance(up_task, Sequence) or not isinstance(down_task, Sequence):
             raise TypeError(
-                'Chain not supported between instances of {up_type} and {down_type}'.format(
+                "Chain not supported between instances of {up_type} and {down_type}".format(
```
Comment (ashb) [inline airflow/models/baseoperator.py]:
```suggestion
                'Chain not supported between instances of {up_type} and {down_type}'.format(
```

Please avoid stylistic only changes

File: airflow/models/baseoperator.py [hunk 3]
```diff
@@ -1566,36 +1567,41 @@ def chain(*tasks: Union[BaseOperator, Sequence[BaseOperator]]):
...
                 )
             )
         up_task_list = up_task
         down_task_list = down_task
         if len(up_task_list) != len(down_task_list):
             raise AirflowException(
-                f'Chain not supported different length Iterable '
-                f'but get {len(up_task_list)} and {len(down_task_list)}'
+                f"Chain not supported different length Iterable "
+                f"but get {len(up_task_list)} and {len(down_task_list)}"
```
Comment (ashb) [inline airflow/models/baseoperator.py]:
```suggestion
                f'Chain not supported different length Iterable '
                f'but get {len(up_task_list)} and {len(down_task_list)}'
```

File: tests/dags/test_chain_xcomargs.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+#
```
Comment (ashb) [inline tests/dags/test_chain_xcomargs.py]: This shouldn't be a new file, but some extra cases with asserts here https://github.com/apache/airflow/blob/fa811057a6ae0fc6c5e4bff1e18971c262a42a4c/tests/models/test_baseoperator.py#L381-L401

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/models/baseoperator.py (hunk 3 above)
Comment (josh-fell) [reply airflow/models/baseoperator.py]: Removed stylistic changes. They were applied via black when saving the file.

File: tests/dags/test_chain_xcomargs.py (hunk 4 above)
Comment (josh-fell) [reply tests/dags/test_chain_xcomargs.py]: Added similar cases for `XComArgs`.

**Comment by uranusjr** [comment]
Should an examples be added to the documentation/docstring as well?

**Comment by josh-fell** [comment]
> Should an examples be added to the documentation/docstring as well?

Great idea. I'll expand the examples in the docstrings for `chain()` and `cross_downstream()`.

File: airflow/models/baseoperator.py [hunk 5]
```diff
@@ -1540,22 +1541,24 @@ def inherits_from_dummy_operator(self):
...
-def chain(*tasks: Union[BaseOperator, Sequence[BaseOperator]]):
+def chain(*tasks: Union[BaseOperator, "XComArg", Sequence[BaseOperator], Sequence["XComArg"]]):
     r"""
     Given a number of tasks, builds a dependency chain.
-    Support mix airflow.models.BaseOperator and List[airflow.models.BaseOperator].
-    If you want to chain between two List[airflow.models.BaseOperator], have to
-    make sure they have same length.
+    Support mix airflow.models.BaseOperator, List[airflow.models.BaseOperator], XComArg, and
+    List[airflow.models.XComArg]. If you want to chain between two List[airflow.models.BaseOperator]
+    or List[airflow.models.XComArg], you have to make sure they have same length.
```
Comment (uranusjr) [inline airflow/models/baseoperator.py]:
```suggestion
    or List[airflow.models.XComArg], you have to make sure they have the same length.
```

**Review by uranusjr — COMMENTED** [review]
One minor comment, overally awesome stuff 😄

File: airflow/models/baseoperator.py (hunk 5 above)
Comment (josh-fell) [reply airflow/models/baseoperator.py]: 👍 Nice catch! Updated.

**Review by uranusjr — APPROVED** [review]

File: airflow/models/baseoperator.py [hunk 6]
```diff
@@ -1595,11 +1652,14 @@ def chain(*tasks: Union[BaseOperator, Sequence[BaseOperator]]):
 
 
 def cross_downstream(
-    from_tasks: Sequence[BaseOperator], to_tasks: Union[BaseOperator, Sequence[BaseOperator]]
+    from_tasks: Union[Sequence[BaseOperator], Sequence["XComArg"]],
+    to_tasks: Union[Sequence[BaseOperator], Sequence["XComArg"]],
```
Comment (ashb) [inline airflow/models/baseoperator.py]: Shouldn't this be

```suggestion
    from_tasks: Sequence[Union[BaseOperator, "XComArg"]],
    to_tasks: Sequence[Union[BaseOperator Sequence["XComArg"]],
```

Comment (ashb) [reply airflow/models/baseoperator.py]: Actually this:

```suggestion
    from_tasks: Sequence[Union[BaseOperator, "XComArg"]],
    to_tasks: Union[BaseOperator, "XComArg", Sequence[Union[BaseOperator, Sequence["XComArg"]]],
```

Comment (uranusjr) [reply airflow/models/baseoperator.py]: Good catch, you’re right, these are heterogeneous lists, not two possible homogeneous lists.

Comment (josh-fell) [reply airflow/models/baseoperator.py]: Agreed, you're right 👍 . Updated.

File: airflow/models/baseoperator.py [hunk 7]
```diff
@@ -1540,22 +1541,24 @@ def inherits_from_dummy_operator(self):
         return getattr(self, '_is_dummy', False)
 
 
-def chain(*tasks: Union[BaseOperator, Sequence[BaseOperator]]):
+def chain(*tasks: Union[BaseOperator, "XComArg", Sequence[BaseOperator], Sequence["XComArg"]]):
```
Comment (ashb) [inline airflow/models/baseoperator.py]:
```suggestion
def chain(*tasks: Union[BaseOperator, "XComArg", Sequence[Union[BaseOperator, "XComArg"]]]):
```

File: airflow/models/baseoperator.py [hunk 8]
```diff
@@ -1540,22 +1541,24 @@ def inherits_from_dummy_operator(self):
...
     r"""
     Given a number of tasks, builds a dependency chain.
-    Support mix airflow.models.BaseOperator and List[airflow.models.BaseOperator].
-    If you want to chain between two List[airflow.models.BaseOperator], have to
-    make sure they have same length.
+    Support mix airflow.models.BaseOperator, List[airflow.models.BaseOperator], XComArg, and
+    List[airflow.models.XComArg]. If you want to chain between two List[airflow.models.BaseOperator]
+    or List[airflow.models.XComArg], you have to make sure they have the same length.
```
Comment (ashb) [inline airflow/models/baseoperator.py]:
```suggestion

    This function accepts values of BaseOperator (aka tasks), XComArg, or lists containing
    either type (or a mix of both in the same list). If you want to chain between two lists you must
    ensure they have the same length.
```

**Review by ashb — APPROVED** [review]

**Pull request merged**